pkg/usecases/authentication/device_code/device_code.go

```diff
@@ -0,0 +1,87 @@
+package device_code
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/int128/kubelogin/pkg/infrastructure/browser"
+	"github.com/int128/kubelogin/pkg/infrastructure/logger"
+	"github.com/int128/kubelogin/pkg/oidc"
+	"github.com/int128/kubelogin/pkg/oidc/client"
+)
+
+type Option struct {
+	SkipOpenBrowser bool
+	BrowserCommand  string
+}
+
+// DeviceCode provides the oauth2 device code flow.
+type DeviceCode struct {
+	Browser browser.Interface
+	Logger  logger.Interface
+}
+
+var errExpired = errors.New("device request expired")
+
+func (u *DeviceCode) Do(ctx context.Context, in *Option, oidcClient client.Interface) (*oidc.TokenSet, error) {
+	u.Logger.V(1).Infof("starting the oauth2 device code flow")
+
+	dr, err := oidcClient.GetDeviceAuthorization(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if dr.VerificationURIComplete == "" {
+		u.Logger.Printf("Please enter the following code when asked in your browser: %s", dr.UserCode)
+		u.openURL(ctx, in, dr.VerificationURI)
+	} else {
+		u.openURL(ctx, in, dr.VerificationURIComplete)
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, time.Duration(dr.ExpiresIn)*time.Second)
+	defer cancel()
+
+	fetchTicker := time.NewTicker(time.Duration(dr.Interval) * time.Second)
+
+	defer fetchTicker.Stop()
+
+	for {
+		select {
+		case <-fetchTicker.C:
+			tokenSet, err := oidcClient.ExchangeDeviceCode(ctx, dr.DeviceCode)
+			if err != nil {
+				u.Logger.V(1).Infof("err:", err)
+				continue
```

Comment:
It looks to treat any errors as acceptable until the timeout has expired or the response is successful, while I believe it should only retry if the response code is "400" and the "error" is "authorization_pending".
See: [Device Access Token Response](https://www.rfc-editor.org/rfc/rfc8628#section-3.5)

> The "authorization_pending" and "slow_down" error codes define
>    particularly unique behavior, as they indicate that the OAuth client
>    should continue to poll the token endpoint by repeating the token
>    request (implementing the precise behavior defined above).  If the
>    client receives an error response with any other error code, it MUST
>    stop polling and SHOULD react accordingly, for example, by displaying
>    an error to the user.

Reply:
I have adjusted the behaviour to only continue on `authorization_pending` and `slow_down` errors, and to slow down by 5 sec if required according to the spec.
Thank you for reviewing and testing!